### PR TITLE
Junction annotation

### PR DESCRIPTION
**Pull request by a-frantz**

### CODE REVIEW

**Comment by a-frantz** [comment]
Do we want to enable a fuzzy search of +-n basepairs? Not sure if that would be desirable, it could mask legitimately novel splices

**Comment by a-frantz** [comment]
Fuzzy matching is now implemented, but is turned off by default. `min-reads` is also disabled by default, but IMO should be enabled to a reasonable value by default. Some testing/input from biologists would be needed to determine what "reasonable" would be for that parameter.

File: .editorconfig [hunk 1]
```diff
@@ -15,7 +15,7 @@ charset = utf-8
 # 4 space indentation
 [*.py]
 indent_style = space
-indent_size = 2
+indent_size = 4
```
Comment (claymcleod) [inline .editorconfig:18]: 🤢

Comment (a-frantz) [reply .editorconfig:18]: Since I switched to `black` this isn't used anymore. And `black` enforces 4 spaces anyway 🤷‍♀️

**Comment by a-frantz** [comment]
@claymcleod @adthrasher I've got `strandedness` runtime back down to the original ~10 seconds with a tabixed GFF. If it's not tabixed, I create a new one (takes ~45 seconds) which can either be reused by name, or detected on reuse.

Planning on going over the changes in code review on Monday.

File: ngsderive/commands/junction_annotation.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+                                str(end),
+                                str(num_reads),
+                                annotation,
+                            ]
+                        ),
+                        file=junction_file,
+                    )
+
+        # if not fuzzy searching, collapsed_junctions is empty and loop is skipped
+        for (intron_start, intron_end), num_reads in sorted(
```
Comment (adthrasher) [inline ngsderive/commands/junction_annotation.py:182]: Why are you looping over the junctions again?

File: ngsderive/commands/junction_annotation.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+    if fuzzy_range == 0:
+        if found_pos in reference_positions:
+            return False, found_pos
+        else:
+            return True, None
+    for pos in reference_positions.irange(
+        found_pos - fuzzy_range, found_pos + fuzzy_range
+    ):  # return first match even if multiple found
+        return False, pos
+    return True, None
```
[the remark this answers is not in the document]
Comment (a-frantz) [reply ngsderive/commands/junction_annotation.py:26]: Sure. I separated it for performance reasons, but they're probably negligible.

File: ngsderive/commands/junction_annotation.py (hunk 2 above)
Comment (a-frantz) [reply ngsderive/commands/junction_annotation.py:182]: This allows read counts of nearby junctions to be added before the `min_reads` filter. So I need to loop through all returned events, collapse the events that fuzzy-match to the same junction, and then loop again for a final filter before writing out.

Comment (adthrasher) [reply ngsderive/commands/junction_annotation.py:178]: Ok, I missed that initially looking at this. Makes sense.

Comment (a-frantz) [reply ngsderive/commands/junction_annotation.py:178]: Just added some comments to make that logic more clear.

**Review by adthrasher — APPROVED** [review]

**Pull request merged**